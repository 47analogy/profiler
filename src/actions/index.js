export function waitingForProfileFromAddon() {
  return {
    type: 'WAITING_FOR_PROFILE_FROM_ADDON'
  };
}

export function receiveProfileFromAddon(profile) {
  return {
    type: 'RECEIVE_PROFILE_FROM_ADDON',
    profile: profile
  };
}

export function requestingSymbolTable(requestedLib) {
  return {
    type: 'REQUESTING_SYMBOL_TABLE',
    requestedLib
  };
}

export function receivedSymbolTableReply(requestedLib) {
  return {
    type: 'RECEIVED_SYMBOL_TABLE_REPLY',
    requestedLib
  };
}

export function startSymbolicating() {
  return {
    type: 'START_SYMBOLICATING'
  };
}

export function profileSymbolicationStep(profile) {
  return {
    type: 'PROFILE_SYMBOLICATION_STEP',
    profile
  };
}

export function doneSymbolicating() {
  return { type: 'DONE_SYMBOLICATING' };
}

export function mergeFunctions(threadIndex, oldFuncToNewFuncMap) {
  return {
    type: 'MERGE_FUNCTIONS',
    threadIndex, oldFuncToNewFuncMap
  };
}

export function assignFunctionNames(threadIndex, funcIndices, funcNames) {
  return {
    type: 'ASSIGN_FUNCTION_NAMES',
    threadIndex, funcIndices, funcNames
  };
}

export function changeSelectedFuncStack(selectedFuncStack) {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack
  };
}
