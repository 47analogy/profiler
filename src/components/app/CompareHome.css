.compareHome {
  width: 75%;
  margin: auto;
  padding: 8em;

  border: 1px solid #CCC;
  border-radius: 3px;
  box-shadow: 0 5px 25px #0b1f50;
  background: #fff;
}

.compareHomeForm {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: 1em;
  align-items: center;
}

.compareHomeSubmitButton {
  grid-column-start: span 2
}

.compareHomeFormLabel {
  white-space: nowrap;
}

@media (max-width: 450px) {
  .compareHome {
    width: auto;
    padding: 1em;
  }
}
