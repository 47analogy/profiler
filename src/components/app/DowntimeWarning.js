/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React, { PureComponent } from 'react';

import { Warning } from '../shared/Warning';

const LOCAL_STORAGE_KEY = 'profileServerMigrationNoticeDismissed2020';

type State = {|
  wasDismissedInThePast: boolean,
|};

export class DowntimeWarning extends PureComponent<{||}, State> {
  state = {
    wasDismissedInThePast: !!localStorage.getItem(LOCAL_STORAGE_KEY),
  };

  _onDowntimeActionClick() {
    window.open('/docs/#/./cloud-migration', '_blank');
  }

  _onClose() {
    localStorage.setItem(LOCAL_STORAGE_KEY, 'dismissed');
  }

  render() {
    // Because this code is planned to be only temporary, let's skip it in tests.
    if (process.env.NODE_ENV === 'test') {
      return null;
    }

    const currentDate = new Date();
    const feb23Date = new Date(Date.UTC(2020, 1, 23));
    if (currentDate >= feb23Date) {
      return null;
    }

    const feb21Date = new Date(Date.UTC(2020, 1, 21));
    if (currentDate < feb21Date && this.state.wasDismissedInThePast) {
      return null;
    }

    return (
      <Warning
        message="The profiler will encounter some downtime on Feb 21 and 22."
        actionText="More information"
        actionTitle="Consult more information about the downtime (opens a new tab)"
        actionOnClick={this._onDowntimeActionClick}
        onClose={this._onClose}
      />
    );
  }
}
