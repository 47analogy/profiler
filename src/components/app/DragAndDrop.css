.dragAndDropArea {
  isolation: isolate;
}

.dragAndDropMessageWrapper {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  align-items: stretch;
  justify-content: center;
  padding: 1em;
  background-color: rgba(255, 255, 255, 0.9);
  opacity: 0;
  pointer-events: none;
  transition: opacity 200ms;
}

.dragAndDropMessageWrapper.dragging {
  opacity: 1;
}

.dragAndDropMessage {
  display: flex;
  width: 100%;
  align-items: center;
  justify-content: center;
  border: 2px dashed #888;
  border-radius: 10px;
  color: #666;
  font-size: 30px;
  text-align: center;
}
