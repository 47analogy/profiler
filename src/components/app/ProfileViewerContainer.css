.ProfileViewerContainer .layout-pane > * {
  box-sizing: border-box;
  width: 100%;
  height: 100%;
}

/* overriding defaults from splitter-layout */
.ProfileViewerContainer > .layout-splitter {
  background: var(--grey-10); /* Same background as sidebars */
  border-left: 1px solid var(--grey-40);
}

.ProfileViewerContainer > .layout-splitter:hover {
  background: var(--grey-40); /* same as the border above */
}


