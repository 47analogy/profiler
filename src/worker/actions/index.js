import { summarizeProfile } from '../../common/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
