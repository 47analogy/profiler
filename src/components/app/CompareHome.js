/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { changeProfilesToCompare } from '../../actions/app';

import explicitConnect from '../../utils/connect';
import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';

import './CompareHome.css';

type DispatchProps = {|
  +changeProfilesToCompare: typeof changeProfilesToCompare,
|};

type Props = ConnectedProps<{||}, {||}, DispatchProps>;

type State = {
  profile1: string,
  profile2: string,
};

class CompareHome extends PureComponent<Props, State> {
  state = { profile1: '', profile2: '' };

  handleInputChange = (event: SyntheticInputEvent<>) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleFormSubmit = (e: SyntheticEvent<>) => {
    e.preventDefault();
    let { profile1, profile2 } = this.state;
    const { changeProfilesToCompare } = this.props;
    if (!profile1.startsWith('http')) {
      profile1 = 'http://' + profile1;
    }
    if (!profile2.startsWith('http')) {
      profile2 = 'http://' + profile2;
    }
    changeProfilesToCompare(profile1, profile2);
  };

  render() {
    const { profile1, profile2 } = this.state;

    return (
      <form className="compareHome" onSubmit={this.handleFormSubmit}>
        <h1>Enter the 2 profile URLs that youâ€™d like to compare</h1>
        <p>
          The tool will extract the data from the selected track and range for
          each profile, and put them both on the same view to make them easy to
          compare.
        </p>
        <label htmlFor="compareHomeProfile1">Profile 1:</label>
        <input
          name="profile1"
          id="compareHomeProfile1"
          className="compareHome-input"
          onChange={this.handleInputChange}
          value={profile1}
        />
        <label htmlFor="compareHomeProfile2">Profile 2:</label>
        <input
          name="profile2"
          id="compareHomeProfile2"
          className="compareHome-input"
          onChange={this.handleInputChange}
          value={profile2}
        />
        <input
          className="compareHome-submit-button"
          type="submit"
          value="Retrieve profiles"
        />
      </form>
    );
  }
}

const options: ExplicitConnectOptions<{||}, {||}, Props> = {
  mapDispatchToProps: { changeProfilesToCompare },
  component: CompareHome,
};

export default explicitConnect(options);
