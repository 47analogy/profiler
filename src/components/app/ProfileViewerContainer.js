/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React from 'react';
import SplitterLayout from 'react-splitter-layout';

import ProfileViewer from './ProfileViewer';
import selectSidebar from '../sidebar';

import { getSelectedTab } from '../../reducers/url-state';
import explicitConnect from '../../utils/connect';

import type { TabSlug } from '../../types/actions';
import type { ExplicitConnectOptions } from '../../utils/connect';

import './ProfileViewerContainer.css';

function dispatchResizeEvent() {
  const event = new UIEvent('resize', { view: window });
  window.dispatchEvent(event);
}

type StateProps = {|
  +selectedTab: TabSlug,
|};

function ProfileViewerContainer({ selectedTab }: StateProps) {
  const Sidebar = selectSidebar(selectedTab);

  /* Note: we use `primaryMinSize` to control the display of the sidebar, instead
   * of `secondaryInitialSize`, because the component SplitterLayout doesn't react
   * to changes to `secondaryInitialSize`.
   * See https://github.com/zesik/react-splitter-layout/issues/14
   */
  return (
    <SplitterLayout
      customClassName="ProfileViewerContainer"
      percentage
      secondaryInitialSize={20}
      onDragEnd={dispatchResizeEvent}
    >
      <ProfileViewer />
      {Sidebar && <Sidebar />}
    </SplitterLayout>
  );
}

const options: ExplicitConnectOptions<{||}, StateProps, {||}> = {
  mapStateToProps: state => ({
    selectedTab: getSelectedTab(state),
  }),
  component: ProfileViewerContainer,
};

export default explicitConnect(options);
