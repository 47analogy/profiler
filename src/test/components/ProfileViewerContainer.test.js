/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { shallowWithStore } from '../fixtures/enzyme';

import ProfileViewerContainer from '../../components/app/ProfileViewerContainer';
import { changeSelectedTab } from '../../actions/app';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';

import type { TabSlug } from '../../types/actions';

describe('app/ProfileViewerContainer', function() {
  const { profile } = getProfileFromTextSamples(`
    A A A
    B B B
    C C H
    D F I
    E E
  `);

  const tabSlugs: TabSlug[] = [
    'stack-chart',
    'marker-chart',
    'flame-graph',
    'marker-table',
    'calltree', // 'calltree' not first on purpose
  ];

  it('renders an initial view', () => {
    const store = storeWithProfile(profile);
    // dive() will shallow-render the wrapped component
    const view = shallowWithStore(<ProfileViewerContainer />, store);
    expect(view.dive()).toMatchSnapshot();
  });

  tabSlugs.forEach((tabSlug: TabSlug) => {
    it(`renders initial views with tab ${tabSlug}`, () => {
      const store = storeWithProfile(profile);
      store.dispatch(changeSelectedTab(tabSlug));

      const view = shallowWithStore(<ProfileViewerContainer />, store);
      expect(view.dive()).toMatchSnapshot();
    });
  });
});
