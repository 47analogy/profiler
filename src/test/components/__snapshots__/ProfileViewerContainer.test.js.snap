// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/ProfileViewerContainer can move between tabs 1`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;

exports[`app/ProfileViewerContainer can move between tabs 2`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;

exports[`app/ProfileViewerContainer can move between tabs 3`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;

exports[`app/ProfileViewerContainer can move between tabs 4`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;

exports[`app/ProfileViewerContainer can move between tabs 5`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={0}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
  <Connect(CallTreeSidebar) />
</t>
`;

exports[`app/ProfileViewerContainer renders an initial view 1`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={0}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
  <Connect(CallTreeSidebar) />
</t>
`;

exports[`app/ProfileViewerContainer renders initial views with tab calltree 1`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={0}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
  <Connect(CallTreeSidebar) />
</t>
`;

exports[`app/ProfileViewerContainer renders initial views with tab flame-graph 1`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;

exports[`app/ProfileViewerContainer renders initial views with tab marker-chart 1`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;

exports[`app/ProfileViewerContainer renders initial views with tab marker-table 1`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;

exports[`app/ProfileViewerContainer renders initial views with tab stack-chart 1`] = `
<t
  customClassName="ProfileViewerContainer"
  onDragEnd={[Function]}
  onDragStart={null}
  onSecondaryPaneSizeChange={null}
  percentage={true}
  primaryIndex={0}
  primaryMinSize={100}
  secondaryInitialSize={20}
  secondaryMinSize={10}
  vertical={false}
>
  <Connect(ProfileViewer) />
</t>
`;
